enhancements/20241211-upgrade-enhancement.md

```diff
@@ -0,0 +1,997 @@
+
+### User Experience in Detail
+
+The upgrade workflow remains consistent with previous versions, with steps outlined below:
+
+1. Administrator checks current cluster version and available upgrades
+   - CLI: `kubectl get settings.harvesterhci.io server-version`
+   - UI: Navigate to the cluster dashboard page and look for the "Version" string
+1. Administrator reviews upgrade prerequisites and release notes
+   - Check support matrix
+   - Review breaking changes and known issues
+   - Verify cluster status with the pre-flight check script
+1. Administrator initiates upgrade process
+   - CLI: `kubectl create -f upgradeplan.yaml`
+   - UI: Click "Upgrade" button, select target version
+1. Administrator monitors upgrade progress
+   - Real-time status in the Upgrade modal on the UI dashboard
+   - CLI: `kubectl get upgradeplans.management.harvesterhci.io <upgradeplan-name> -w`
+1. Administrator validates post-upgrade functionality
+   - Verify all VMs are running
+   - Check storage and networking
+   - Confirm new features are available and bug fixes are applied
+1. Administrator handles any issues
+   - Access to upgrade logs
+   - Start another round of upgrade after the cluster becomes stable
+   - Support contact information
+
+### API Changes
+
+Introducing new APIs: **UpgradePlan** and **Version**. This new UpgradePlan CRD (Custom Resource Definition) is designed to work closely with the Version CRD. The Version CRD is almost identical to the previous version, as the surrounding functions do not change significantly. The existing Upgrade Responder mechanic can be preserved, albeit with a slight change to the CRD API group. Both CRDs will be under the `management.harvesterhci.io` API group and be cluster-scoped. Being cluster-scoped not only fits the semantics as both APIs are relevant to upgrading the entire cluster, but not just for any specific component, but also provides a better mechanism when it comes to resource purge. In Upgrade V1, both CRDs are namespaced-scope, which results in the inability to cascade delete the owned resources that reside in namespaces other than `harvester-system`. We had no choice but to write dedicated code to actively wipe them out when deletion happens to Upgrade CRs.
+
+The UpgradePlan CRD looks like the following:
+
+```go
+const (
+	// Available means the UpgradePlan is ready to be reconciled.
+	UpgradePlanAvailable string = "Available"
+	// Progressing means the cluster is currently applying the UpgradePlan.
+	UpgradePlanProgressing string = "Progressing"
+	// Degraded means the progress of the upgrade is stalled due to issues.
+	UpgradePlanDegraded string = "Degraded"
+)
+
+const (
+	NodeStateImagePreloading     string = "ImagePreloading"
+	NodeStateImagePreloaded      string = "ImagePreloaded"
+	NodeStateKubernetesUpgrading string = "KubernetesUpgrading"
```

Comment:
Do we need another status for the charts?
Or we unify it on `KubernetesUpgrading`?